src/common/sv2_noise.h

```diff
@@ -0,0 +1,298 @@
+
+    /* The certificate serializes pubkeys in x-only format, not EllSwift. */
+    XOnlyPubKey m_static_key = {};
+
+    [[nodiscard]] bool Validate(XOnlyPubKey authority_key);
+
+    template <typename Stream>
+    // The static_key is signed for, but not serialized.
+    void Serialize(Stream& s) const
+    {
+        s << m_version
+          << m_valid_from
+          << m_valid_to
+          << m_sig;
+    }
+    template <typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> m_version
+          >> m_valid_from
+          >> m_valid_to
+          >> m_sig;
+    }
+};
+
+/*
+ * The CipherState uses m_key (k) and m_nonce (n) to encrypt and decrypt ciphertexts.
+ * During the handshake phase each party has a single CipherState, but during
+ * the transport phase each party has two CipherState objects: one for sending,
+ * and one for receiving.
+ *
+ * See chapter "5. Processing rules" of the Noise paper.
+ */
+class Sv2CipherState
+{
+public:
+    static constexpr size_t HASHLEN{32};
+
+    Sv2CipherState() = default;
+    explicit Sv2CipherState(uint8_t key[HASHLEN]);
+
+    /** Decrypt message
+     * @param[in] associated_data associated data
+     * @param[in] ciphertext message with encrypted and authenticated chunks.
+     * @param[out] plain message (defragmented)
+     * @returns whether decryption succeeded
+     */
+    [[nodiscard]] bool DecryptWithAd(Span<const std::byte> associated_data, Span<std::byte> ciphertext, Span<std::byte> plain);
+
+    /** Encrypt message
+     * @param[in] associated_data associated data
+     * @param[in] plain message
+     * @param[out] ciphertext message with encrypted and authenticated chunks.
+     * @returns whether decryption succeeded
+     */
+    [[nodiscard]] bool EncryptWithAd(Span<const std::byte> associated_data, Span<const std::byte> plain, Span<std::byte> ciphertext);
+
+    /** The message will be chunked in NOISE_MAX_CHUNK_SIZE parts and expanded
+     *  by 16 bytes per chunk for its MAC.
+     *
+     * @param[in] plain message. Can't point to the same memory location as ciphertext,
+     *                  because each encrypted message chunk would override the
+     *                  start of the next plain text chunk.
+     * @param[out] ciphertext   message with encrypted and authenticated chunks
+     * @return whether encryption succeeded. Only fails if nonce is uint64_max.
+     */
+    [[nodiscard]] bool EncryptMessage(Span<const std::byte> plain, Span<std::byte> ciphertext);
+
+    /** Decrypt message.
+     *
+     * @param[in] ciphertext encrypted message
+     * @param[out] plain decrypted message. May point to the same memory location
+     *                  as ciphertext. The result is defragmented.
+     */
+    [[nodiscard]] bool DecryptMessage(Span<std::byte> ciphertext, Span<std::byte> plain);
+
+private:
+    uint8_t m_key[HASHLEN];
```

Comment:
Cpp Core Guidelines explicitly want us to avoid C-Arrays [SL.Con.1](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#slcon1-prefer-using-stl-array-or-vector-instead-of-a-c-array) and prefer STL's`std::array` which does not degenerate to a pointer and knows its size. Would it be better for us here to use `std::array<uint8_t, HASHLEN> m_key` or it is not justified in our context?

Reply:
Okay

Reply:
@itornaza thanks, I'll look into your patch. Another guiding principle would be to stay close to `bip324.h` in terms of coding style. I haven't compared with that yet.